### DISCUSSION TITLE
Temperature probes

### DISCUSSION THREAD

**Discussion by Bournetoride** [Beormund/Tasmota32-Multi-Zone-Heating-Controller#9] [General]
1. Can I use DS18B20 probes as room temperature input.
2. My ultimate aim is to use this fantastic project as the basis for my solar water heating controller.
This has 3 probes:-
top of tank, solar manifold, bottom of tank
in the original setup once the manifold exceeded top of tank by 5'C then the pump starts to take water from bottom of tank out to the manifold, bring the heated water back into top of tank, with a small hysteresis
my first modification was to change the setpoint slowly during the afternoon (1pm -> sunset) from top of tank +5' down to bottom of tank +10'
Now I'm trying to Tasmotise it and Berry scripts look like the way forward and your project looks like the best basis, using a ESP32C3 board, I have the probes connected & reading but "room temp":null is in the console and docs seem to suggest I need to send the temp in by MQTT 
3. Once i have the room temp (manifold) then I will work out how to adjust setpoint according to tank temps & time of day

**Comment by Beormund** [comment]
Apologies - only just seen this.  With regards to point 1). I don't see why not. There are a couple of ways of doing this. You can control a heating zone's relay by using the command heatingzone<x> heat, so for zone 1: heatingzone1 heat to turn on the relay and heatingzone1 idle to turn off the relay. Or, you can enable the thermostat support option and feed temperatures in like so:

To update the target or room temperature (together or individually):
HeatingZone<x> {"update": {"target": 22.5, "room": 21.3}}
HeatingZone<x> {"update": {"target": 22}}
HeatingZone<x> {"update": {"room": 21}}
NB: The THERM option needs to be enabled via the web UI or the HeatingOptions command. 

I have yet to implement a basis hysteresis algorithm though so currently the heating curve is inefficient.

Looks like an interesting project! Please do keep me posted on how you get on.

**Reply by Bournetoride** [reply]
The temperature probes will be attached to the same ESP as the controller, so the update is not done from MQTT, it will come from reading DS18B20-1 & DS18B20-2.
Looking deeper maybe I need to pass temperatures in using TasmotaGlobal user_globals
then in xdrv_52_3_berry_tasmota.ino add "get user" pointing to those globals

**Reply by Beormund** [reply]
> The temperature probes will be attached to the same ESP as the controller, so the update is not done from MQTT, it will come from reading DS18B20-1 & DS18B20-2. 

The Heating Controller commands are Tasmota commands not MQTT commands so you can execute them in all the usual ways: using the console, using berry: tasmota.cmd('HeatingZone {"update": {"room": 21}}'), MQTT, or via web requests etc. Why not use berry to read the sensors and then issue the commands to the controller? 

https://tasmota.github.io/docs/Commands/#how-to-use-commands

**Reply by Bournetoride** [reply]
> Why not use berry to read the sensors and then issue the commands to the controller?

Mainly because there isn't a "berried" DS18B20 driver and I have no idea how to do it myself, its not a simple device

**Comment by Beormund** [comment]
The Tasmota driver for DS18B20 has standard rule triggers so you could use berry's tasmota.add_rule to capture the temperature events.

https://tasmota.github.io/docs/DS18x20/
https://tasmota.github.io/docs/Berry/#rules

**Comment by Bournetoride** [comment]
OK, struggling a bit with the rule call, got this code in
26 var DS18B20_1 
    var DS18B20_2 
    var DS18B20_3 
.
.
390     #Get DS18B20 temperatures
    def rule_dsb(value, trigger)
        var i=string.find(trigger,"DS18B20-")
        var tr=string.split(trigger,i+46)   # DS18B20-1":{"Id":"0218917752DC","Temperature":17.9
        var dsnum=number.split(trigger,i+8)
        var temp=number(tr[4])
        var DS18B20_(dsnum) = temp
    end

    tasmota.add_rule("DS18B20-1",rule_dsb)
    tasmota.add_rule("DS18B20-2",rule_dsb)
    tasmota.add_rule("DS18B20-3",rule_dsb)

console says
00:00:00.413 BRY: failed to run compiled code 'syntax_error' - /heating.be:399: unexpected token 'tasmota'

**Reply by chivas03** [reply]
Do you already have a solution to integrate ds18b20 sensor? I want to make a similar device - Having 4 temperature inputs and 4 output relays on an ESP32. Alternatively, I could make it with a tasmota ZB bridge using websend to transmit the temperature values to the ESP32. I need to try if I can send the values to the berry script

**Reply by chivas03** [reply]
Today I tried to push the temperatures to the multi-zone-heating controller with DS18B20 probes. In my test set-up I used 3 sensors and it worked fine. Here is the rule:

Rule1
ON DS18B20-1#Temperature DO HeatingZone1 {"update": {"room": %value%}} ENDON
ON DS18B20-2#Temperature DO HeatingZone2 {"update": {"room": %value%}} ENDON
ON DS18B20-3#Temperature DO HeatingZone3 {"update": {"room": %value%}} ENDON

Similar set-up would work with a websend task from the Tasmota ZB bridge without Home assistant or IObroker. With websend I would write the specific temperature value into a variable.
